Use FixedString for PostgreSQL's types that have a fixed length

pamarcos opened the pull request:
### Changelog category (leave one):
- Improvement

Use FixedString for PostgreSQL's CHARACTER, CHAR and BPCHAR

Close https://github.com/ClickHouse/ClickHouse/issues/77303

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use FixedString for PostgreSQL's CHARACTER, CHAR and BPCHAR

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

https://github.com/ClickHouse/clickhouse-docs/pull/3489

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

pamarcos committed:
Use FixedString for PostgreSQL's types that have a fixed length

Algunenano commented on src/Databases/PostgreSQL/fetchPostgreSQLTableStructure.cpp:
```diff
@@ -135,8 +136,27 @@ static DataTypePtr convertPostgreSQLDataType(String & type, Fn<void()> auto && r
     if (!res)
-        res = std::make_shared<DataTypeString>();
+    {
+        if (type.ends_with(')'))
```
`geometry(Point, 4326)` would like a word

Algunenano replied:
We need a better way than using parenthesis

pamarcos replied:
I'll leave it then to the original implementation that I had where I was checking that the type started with `character`. I thought moving it down here would make it more flexible for some other way of specifying strings with particular sizes, but I see it could bring more trouble than anything.

This is the commit [`050f74a` (#77304)](https://github.com/ClickHouse/ClickHouse/pull/77304/commits/050f74a85fe6c8c1dd331e64d3694d5269edcaae)

Algunenano replied:
But this is again not correct -> https://www.postgresql.org/docs/current/datatype-character.html

`character(n), char(n), bpchar(n)` are fixed strings.
`character varying(n), varchar(n)` are variable length with limit (so they should not be matched to CH's FixedString).

Let's hardcode good assumptions instead of using conditionals

pamarcos replied:
>character(n), char(n), bpchar(n) are fixed strings.
>Let's hardcode good assumptions instead of using conditionals

Roger, all those 3 can be filtered by checking if the type starts with `character(`.

pamarcos committed:
Keep the fixed-size String map to something more conservative

pamarcos committed:
Only convert known fixed-sized strings

https://www.postgresql.org/docs/current/datatype-character.html

pamarcos committed:
Merge branch 'master' into add-fixedstring-to-postgresql-function-table-conversion

pamarcos committed:
Add tests to ensure conversion of FixedString works for PostgreSQL

Algunenano reviewed (commented):
Can we please add tests, both ways, for the conversions with incomplete / padded strings? It seems padding will be different and might affect results.

pamarcos commented:
> Can we please add tests, both ways, for the conversions with incomplete / padded strings? It seems padding will be different and might affect results.

Sure. I also noticed my changes broke [test_postgresql_replica_database_engine_1/test.py::test_different_data_types](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=77304&sha=e8be6a98f164be03e90f863df3df979a0a103b7f&name_0=PR&name_1=Integration+tests+%28release%2C+4%2F4%29&name_2=test_postgresql_replica_database_engine_1%2Ftest.py%3A%3Atest_different_data_types). I'm also checking that out

pamarcos committed:
Add test for PostgreSQL protocol for FixedString

pamarcos committed:
Fix FixedString conversion from PostgreSQL to ClickHouse

pamarcos committed:
More elegant fix for FixedString conversion to PostgreSQL

pamarcos committed:
Merge branch 'master' into add-fixedstring-to-postgresql-function-table-conversion

pamarcos committed:
Fix style

pamarcos committed:
Add FixedString support to StorageMaterializedPostgreSQL

pamarcos committed:
Merge branch 'master' into add-fixedstring-to-postgresql-function-table-conversion

pamarcos committed:
Fix include

pamarcos committed:
Merge branch 'master' into add-fixedstring-to-postgresql-function-table-conversion

pamarcos committed:
Add comment

pamarcos committed:
Merge branch 'master' into add-fixedstring-to-postgresql-function-table-conversion

pamarcos commented on src/Core/PostgreSQLProtocol.cpp:
```diff
@@ -28,6 +28,8 @@ ColumnTypeSpec convertTypeIndexToPostgresColumnTypeSpec(TypeIndex type_index)
             return {ColumnType::FLOAT8, 8};
 
         case TypeIndex::FixedString:
+            return {ColumnType::BPCHAR, -1};
```
This is not strictly needed for this particular PR, but I thought we'd rather have the proper `ColumnType` anyways

pamarcos committed:
Another minor comment

pamarcos commented on src/Storages/StoragePostgreSQL.cpp:
```diff
@@ -216,6 +221,63 @@ void StoragePostgreSQL::read(
+    {
+        auto n = assert_cast<const DataTypeFixedString &>(*non_nullable_type).getN();
+        auto custom_desc = std::make_unique<DataTypeCustomDesc>(std::make_unique<DataTypeCustomFixedName>(fmt::format("FixedString({})", n)), std::make_shared<SerializationPostgreSQLFixedString>(n));
+        DataTypeFactory::instance().setCustom(data_type, std::move(custom_desc));
```
Changing the serialization for this type is the less ugly way and the best abstraction that I found to workaround this. In order to make it work, I had to create a new `setCustom` to in `DataTypeFactory` because it's a private method of `IDataType`

Algunenano commented on src/DataTypes/DataTypeFactory.h:
```diff
@@ -35,6 +35,7 @@ class DataTypeFactory final : private boost::noncopyable, public IFactoryWithAli
     DataTypePtr get(const ASTPtr & ast) const;
     DataTypePtr getCustom(DataTypeCustomDescPtr customization) const;
     DataTypePtr getCustom(const String & base_name, DataTypeCustomDescPtr customization) const;
+    void setCustom(DataTypePtr & type, DataTypeCustomDescPtr customization) const;
```
This is extremely confusing for 2 reasons:
* It's only using DataTypeFactory because it's a friend of IDataType.
* It could be static. It's not modifying and does not need the global DataTypeFactory instance.

pamarcos replied:
Alright, I agree it can be confusing. Fixed in [`8c0a5ad` (#77304)](https://github.com/ClickHouse/ClickHouse/pull/77304/commits/8c0a5add179432f65187e77497f14a2b3389e1b0)

But I guess the same could be said for `getCustom`->`get`->`getImpl`->`tryParseQuery`/`parseQuery`

Algunenano commented on src/Storages/StoragePostgreSQL.cpp:
```diff
@@ -216,6 +221,63 @@ void StoragePostgreSQL::read(
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't serialize type {} as a FixedString", nested_column->getName());
+
+        /// Write only the amount of UTF-8 characters actually used. PostgreSQL complains if we pass
```
What will happen if the FixedString contains '\0' naturally? For example `'1\03'::FixedString(3)`

pamarcos replied:
That's a good point. PostgreSQL's strings do not accept intermediate \0 for strings, so we cannot really do a 1:1 conversion. We have 2 options here: either we throw a failure because of that:

```
pqxx::data_exception: Failure during '[END COPY]': ERROR:  invalid byte sequence for encoding "UTF8": 0x00
```

Or we convert any intermediate `\0` to a whitespace, which is what I did at [`c9bf83a` (#77304)](https://github.com/ClickHouse/ClickHouse/pull/77304/commits/c9bf83af398e403531d60f9fab6f22abbaa4bf8a)

pamarcos committed:
Merge branch 'master' into add-fixedstring-to-postgresql-function-table-conversion

pamarcos committed:
Convert to static as per PR feedback

It's true that it doesn't need anything from the DataTypeFactory
instance, but AFAIU it's the same for the getCustom methods.

pamarcos committed:
Replace intermediate \0 with whitespace as a trade-off
